[게시판 회원 기능] 에헴 미션 제출합니다.

hyogenie-v opened the pull request:
안녕하세요 에헴입니다!

고민할 부분이 많아 신경쓰다보니 일요일 밤늦게 미션을 제출하게 되었습니다.
혹시 귀한 주말 시간을 빼앗는 건 아닌지 죄송스러운 마음입니다.....
다음 미션을 진행할 때는 완전하게 마음에 들지 않더라도 제출 기한을 정해놓고 그에 맞추는 연습을 해야겠습니다😢

미션을 진행하면서 가장 어려우면서 고민을 많이 한 부분은 테스트 코드입니다.
ATDD는 아직 헷갈리는 부분이 많아서 여기저기에서 수집한 정보를 참고해 기준을 세워서 작성해봤습니다.
하지만 아직 마음에 들지 않아 고치려다 못한 부분들이 좀 있어서 그것들을 봐주셨으면 합니다!
1. 테스트에서 회원가입이나 로그인처럼 **공통적으로 선행되어야 하는 단계가** 있는데 그 부분을 깔끔하게 고치고 싶습니다.
`@MockBean` 어노테이션을 사용하면 Bean 객체의 상태(회원가입 등)들을 의도적으로 관리할 수 있다고 해서 나름대로 열심히 찾아보고 따라해보았는데 성공하지 못했습니다.(ㅠㅠ)
코치님께서 예시를 들어주시거나 참고 링크를 알려주시면 제가 피드백을 반영할 때 큰 도움이 될 것 같습니다!
2. 개별 테스트는 통과하는데 한 번에 돌리면 깨지는 테스트가 많습니다.
예를 들어서, 중복된 이메일을 가입할 수 없도록 예외를 걸어놓으니 다른 테스트에서 같은 객체를 추가하려고 하면 실패하는 경우가 있습니다.
원래는 다같이 돌려도 테스트가 모두 성공하도록 각 테스트마다 필요한 엔티티 객체의 상태를 다르게 줬었는데, 그것보다는 중복코드가 제거된 상태로 둔 다음 방법을 찾아서 적용하는 게 낫다고 판단해서 그대로 두었습니다.
찾아보니 `@RunWith` 어노테이션을 사용하면 해결할 수 있다고 했는데 맞는 방법인가요? 저는 사용법을 잘 몰라서 그런지 이것도 실패해서요....
설명을 해주시면 정말 감사하겠습니당ㅎㅎ

세세한 리뷰 부탁드립니다!~

indiveloper committed:
docs README.md 작성

indiveloper committed:
feat : 게시물의 정보를 담고 있는 bean 객체인 article class 구현

indiveloper committed:
docs: README.md 구현 기능 추가

soojinroh committed:
Update README.md

soojinroh committed:
Update README.md

soojinroh committed:
Update README.md

indiveloper committed:
feat 정적 파일 수정 시, 자동 재시작 및 LiveReload 설정

indiveloper committed:
feat 새로운 ArticleId를 생성하는 기능 구현

List<Article>에서 가장 큰 articleId보다 1만큼 큰 Long 타입 값을 반환

indiveloper committed:
fix: Article 클래스 수정

1. Article 클래스를 POJO로 수정
2. 불필요한 필드 삭제

indiveloper committed:
fix: articleId에 해당하는 Article 객체 조회 메서드 수정

indiveloper committed:
feat: 게시글 상세 조회 기능 구현

1. articleId에 해당하는 게시글 조회 메서드 생성
2. 게시글 상세 조회 웹 UI 수정

indiveloper committed:
feat: 게시글 상세 조회 기능 구현

1. articleId에 해당하는 게시글 상세 조회 메서드 구현
2. articleId에 해당하는 게시글을 가져와 Web UI에 구현

indiveloper committed:
feat: 웹 UI 수정

1. index.html, article.html, article-edit.html의 정적 파일 경로 수정
2. 전달받은 데이터를 타임리프를 이용하여 동적으로 출력하는 기능 적용

indiveloper committed:
feat : 게시물 수정 기능 구현

1. 수정하려는 게시글의 정보를 article-edit.html에 출력
2. 전달받은 Article 객체와 동일한 articleId를 가진 Article 객체를 List<Article>에서 바꾸는 메서드 구현
3. coverURL이 입력되지 않은 경우, 기본 이미지 URL 값을 할당하도록 수정

indiveloper committed:
refactor: 정적 파일의 경로를 수정

indiveloper committed:
refactor: 불필요한 파일 삭제

indiveloper committed:
feat : 환경설정 변경

indiveloper committed:
docs : README.md 수정

indiveloper committed:
feat : add lovely sloth image file

live like sloth, eat like sloth, sleep like sloth

indiveloper committed:
refactor : update REST 메서드 변경

1. post 방식의 메서드를 put으로 변경

indiveloper committed:
feat : 게시글 삭제 기능 구현

1. delete 메서드 방식을 이용해 게시글 삭제
2. article.html에 삭제 버튼 추가

indiveloper committed:
feat : 전체 게시글 목록 조회 기능 구현

1. get 메서드 방식을 이용해 전체 게시글 조회 메서드 구현
2. index.html에 전체 글 목록 출력

indiveloper committed:
refactor: css 오류 수정 및 게시글 상세 조회 페이지의 링크 경로 수정

indiveloper committed:
fix : fix conflict

indiveloper committed:
feat: ArticleRepository 테스트 추가

indiveloper committed:
refactor: ArticleController 클래스 테스트 리팩토링

1. 매직 넘버 상수화
2. response의 상태코드 확인 메서드 수정(is3xxRedirection() -> isFound())

indiveloper committed:
refactor : html 중복 제거

hyogenie-v committed:
fix: change port

hyogenie-v committed:
fix: 그래들 라이브러리 의존성 옵션 변경

hyogenie-v committed:
refactor: ArticleRepository 클래스가 articleId를 부여하는 로직 변경

hyogenie-v committed:
refactor: URI 컨벤션 수정

1. 웹 UI form 태그의 action 경로 수정
2. ArticleController 클래스의 URI 매핑 파라미터가 자원의 행위를 나타내지 않도록 수정
3. 테스트 코드 수정

hyogenie-v committed:
refactor: 중복되는 html 코드 추출 및 include

1. 중복되는 웹 UI의 html 코드를 추출하여 파일을 include하도록 수정한다
2. ArticleController 테스트 클래스에서 상태 코드도 함께 확인하도록 수정한다

hyogenie-v committed:
refactor: 중복 테스트 코드를 setup() 메서드로 추출

hyogenie-v committed:
refactor: 컨트롤러 메서드의 리턴타입 수정

hyogenie-v committed:
refactor: 컨트롤러에서 비즈니스 로직 제거

hyogenie-v committed:
refactor: Article 클래스의 getter 사용 줄임

1. Article 클래스에 hasSameArticleId() 메서드를 추가하여 메시지를 보내도록 수정한다
2. ArticleRepository 클래스의 findArticleIndexByArticle 메서드를 스트림으로 변경한다

hyogenie-v committed:
refactor: 모델 애트리뷰트의 변경 방식을 setter에서 생성자 주입 방식으로 변경

1. 뷰로부터 전달받은 데이터를 ArticleDto 클래스가 받도록 수정한다
2. 수정된 게시글의 데이터를 Article 클래스 내부에서 적용하도록 수정한다

hyogenie-v committed:
refactor: 사용하지 않는 메서드(generateNewId()) 삭제

hyogenie-v committed:
feat: WebTestClient를 사용한 새로운_게시글_추가 API 테스트 추가

hyogenie-v committed:
docs: README.md 작성

hyogenie-v committed:
feat: 그래들 의존성 추가

hyogenie-v committed:
feat: 웹 UI 중복 제거 및 thymeleaf 적용

hyogenie-v committed:
feat: Article 클래스 구현

hyogenie-v committed:
feat: ArticleRepository의 행위를 하는 기능 구현

1. CrudRepository 인터페이스를 상속하는 ArticleRepository 인터페이스 선언
2. Article 클래스의 생성자 추가
3. ArticleRepository의 게시글 추가, 수정, 삭제에 대한 테스트 작성

hyogenie-v committed:
feat: ArticleDto 클래스 추가

hyogenie-v committed:
feat: Article 클래스와 ArticleDto 클래스를 변환하는 ArticleAssembler 클래스 추가

1. Article 객체를 ArticleDto 객체로 변환하는 메서드 추가
2. ArticleDto 객체를 Article 객체로 변환하는 메서드 추가

hyogenie-v committed:
feat: ArticleController 클래스에 index 페이지 매핑 기능 구현

hyogenie-v committed:
refactor: 인수테스트 수정

1. ArticleControllerTest에서 ArticleRepository 객체를 사용하지 않고 인수테스트를 진행하도록 수정한다
2. ArticleRepository의 메서드를 간결하게 수정한다

hyogenie-v committed:
refactor: 컨트롤러의 Article 객체를 setter가 아닌 생성자로 데이터를 주입받도록 변경

1. Article 클래스에 생성자를 선언한다
2. Article 클래스의 setter를 삭제한다
3. 사용되지 않는 ArticleDto 클래스를 삭제한다
4. 테스트를 수정한다

choihz committed:
refactor: 사용하지 않는 생성자 삭제

choihz committed:
fix: URL 경로 및 오류 수정

choihz committed:
refactor: ArticleRepository 클래스 수정

1. ArticleRepository 클래스가 CrudRepository가 아닌 JpaRepository를 extends하도록 수정
2. ArticleRepository 클래스의 모든 메서드를 정적 메서드로 수정

choihz committed:
feat: Controller와 Service를 분리

1. Controller의 ArticleRepository 접근 코드를 Service로 분리
2. 테스트 코드 수정

hyogenie-v committed:
feat: 모든 게시글 조회 기능 구현

hyogenie-v committed:
feat: 게시글 수정 기능 구현

hyogenie-v committed:
feat: 게시글 삭제 기능 구현

hyogenie-v committed:
feat: 회원가입 및 로그인 페이지 매핑 구현

hyogenie-v committed:
feat: 웹 UI의 경로 재설정

hyogenie-v committed:
feat: 사용자의 정보를 저장하는 User 클래스 생성

hyogenie-v committed:
refactor: 패키지 구조 변경

hyogenie-v committed:
feat: 회원가입에 관여하는 모든 클래스 생성

hyogenie-v committed:
feat: 회원 등록 기능 구현

1. 중복된 이메일을 입력했는지 확인하고, 중복인 경우 예외를 던진다
2. 새로운 회원을 UserRepository에 등록한다
3. UserRepository 인터페이스에 이메일을 검색하는 메서드를 등록한다
4. 테스트 코드를 작성한다

hyogenie-v committed:
feat: 회원가입 기능 구현

hyogenie-v committed:
feat: 회원가입 규칙 적용

hyogenie-v committed:
feat: 회원을 조회하는 기능 구현

1. User 객체와 UserDto 객체를 변환하는 UserAssembler 클래스의 writeDtos 메서드 구현
2. UserService 클래스에서 등록된 모든 회원을 리턴하는 getAllUsers 메서드 구현
3. UserRepository 객체에 등록된 모든 회원을 조회하는 테스트 작성

hyogenie-v committed:
feat: 웹 UI 회원 리스트 조회 기능 구현

hyogenie-v committed:
refactor: 중복된 웹 UI 제거

hyogenie-v committed:
feat: 로그인 기능 구현

1. UserRepository 인터페이스에 countByEmail 메서드와 findUserByEmailAndPassword 메서드 추가
2. 로그인시 입력받은 이메일과 비밀번호가 존재하거나 일치하는지 확인
3. 일치하는 회원 정보로 로그인시 세션 등록
4. 로그인된 헤더에 회원 이름 출력
5. 테스트 코드 작성

hyogenie-v committed:
refactor: 웹 UI 파일명 수정

choihz committed:
feat: 마이페이지 및 정보 수정 기본 구조

1. UserController에 마이페이지 및 정보수정 페이지 매핑
2. 웹 UI의 경로 수정
3. 웹 UI에 회원 탈퇴 버튼 추가

choihz committed:
feat: 로그아웃 기능 구현

1. 로그아웃 기능 구현
2. 클래스별 GeneratedValue를 갖도록 strategy에 IDENTITY 설정

choihz committed:
feat: 마이페이지에 회원 정보 출력 기능 구현

1. 마이페이지에서 로그인된 회원의 정보를 출력한다
2. 로그인 되어 있지 않은 상태에서 접속하면 로그인 페이지로 리다이렉트한다

choihz committed:
feat: 회원 정보 수정 페이지에 정보 출력 기능 구현

1. 회원 정보 수정 페이지에서 로그인된 회원의 정보를 출력한다
2. 로그인 되어 있지 않은 상태에서 접속하면 로그인 페이지로 리다이렉트한다

choihz committed:
feat: 회원 정보 수정 기능 구현

1. UserController 클래스에서 매핑 및 세션을 수정하는 메서드 구현
2. UserService 클래스에서 회원의 정보를 수정하는 메서드 구현
3. mypage-edit.html에 form 태그 추가

choihz committed:
feat: 회원 탈퇴 기능 구현

1. 회원 정보 수정 페이지에서 회원 탈퇴 버튼을 클릭하면 /mypage/delete로 리다이렉트한다
2. UserController 클래스에 경로 매핑 메서드를 구현한다
3. UserService 클래스에서 회원을 삭제하는 메서드를 구현한다
4. UserRepositoryTest에 테스트 메서드를 작성한다

choihz committed:
feat: User 객체 생성시 각 데이터 검증

1. User 객체에 올바르지 않은 데이터가 전달되면 예외가 발생한다
2. 테스트 코드 작성
3. InvalidUserDataException 클래스 추가

choihz committed:
feat: 회원가입 시 예외에 대한 뷰 출력 기능 구현

choihz committed:
feat: 회원가입 시 이메일 중복 예외에 대한 뷰 출력 기능 구현

choihz committed:
feat: 로그인 시 발생하는 예외 처리 적용

1. 로그인 시 발생하는 예외를 뷰에 출력
2. 부적절한 메서드명 수정
3. FailedLoginException 클래스 추가

hyogenie-v committed:
refactor: signup.html에서 불필요한 주석 제거

hyogenie-v committed:
refactor: UserController 및 UserService 클래스에서 로그인 관련 기능 분리

hyogenie-v committed:
refactor: LoginController 클래스에 로그인 예외 처리 기능 추가

hyogenie-v committed:
refactor: 회원가입 시 전달받은 입력값에 대한 검증 처리 로직 수정

1. UserDto 클래스에 @Valid를 추가한다
2. User 클래스에서 입력값 검증 메서드를 삭제한다
3. 회원 저장 시 저장된 User 객체의 정보를 찍는 로그를 추가한다
4. UserControllerTest에 유효하지 않은 정보를 입력받았을 경우에 대한 예외처리 테스트를 추가한다

hyogenie-v committed:
feat: UserDto 객체의 각 데이터가 NotNull이도록 수정

hyogenie-v committed:
refactor: 로그인하지 않은 사용자가 특정 페이지 접속 시, ExceptionHandler를 통해 로그인 페이지가 보여지도록 변경

1. UserController에서 UnknownHostAccessException 발생 시 ExceptionHandler로 던져지도록 변경한다
2. UnknownHostAccessException 익셉션 클래스를 추가한다
3. 로그인 뷰가 에러 메시지를 받으면 alert로 출력되도록 수정한다

hyogenie-v committed:
feat: signup 및 mypage-edit 뷰의 input에 pattern과 placeholder값 수정

hyogenie-v committed:
UserService 클래스에 존재하지 않는 회원의 정보를 수정하는 경우에 대한 예외 추가

hyogenie-v committed:
refactor: 탈퇴하려는 회원 검색 기준을 email이 아닌 userId를 기준으로 수정

1. 회원 탈퇴 URI를 변경한다
2. UserService 클래스의 deleteUser 메서드가 email이 아닌 userId를 파라미터로 전달받도록 수정한다

hyogenie-v committed:
feat: 조건에 맞지 않는 입력값으로 회원 정보 수정하는 경우에 대한 예외처리 추가

hyogenie-v committed:
test: UserController에 대한 테스트 추가

hyogenie-v committed:
test: ArticleController에서 중복되는 코드를 메서드로 추출

hyogenie-v committed:
test: UserService 테스트에서 중복되는 코드 제거

1. UserServiceTest에서 중복되는 코드를 멤버변수와 메서드로 추출한다
2. DuplicateEmailException 클래스의 이름을 사용 의미에 맞게 DuplicateUserException으로 변경한다

hyogenie-v committed:
test: LoginContoller 테스트 메서드에서 중복되는 코드를 메서드로 추출

hyogenie-v committed:
test: UserRepository 테스트에 추가된 메서드에 대한 테스트 추가

hyogenie-v committed:
refactor: 컨벤션 포매팅 및 사용하지 않는 메서드 삭제'

hyogenie-v committed:
chore: mysql 연동 그래들 추가

hyogenie-v committed:
Merge branch 'hyojin2' into dasistHYOJIN

hyogenie-v committed:
fix: UserDto 클래스의 이름 유효성 검사 정규식 수정

hyogenie-v committed:
Merge branch 'dasistHYOJIN' of https://github.com/dasistHYOJIN/jwp-blog into dasistHYOJIN

hyogenie-v committed:
Merge remote-tracking branch 'upstream/dasistHYOJIN' into dasistHYOJIN

gracefulBrown reviewed (changes requested):
안녕하세요~ 리뷰어 브라운입니다 :)
회원기능구현 하시느라 고생많으셨습니다!
우선 아직 강의에서 다루지 않은 mock테스트에 대해서는 혼동이 올 수 있을것 같아요.
이 부분에 대해서는 오프라인 리뷰를 통해 간단히 알려드리고
강의에서 자세한 내용과 사용법에 대해서 알려드리겠습니다.
코멘트 남긴 부분에 대해서 고민해보시고 적용해보면 좋을 것 같아요~
궁금한 점 있으면 언제든 질문주시거나 DM주세요 ^^

hyogenie-v committed:
refactor: 잘못된 메서드명 수정

1. UserService의 findUserByEmailAndPassword() 메서드의 이름을 findUserVyEmail()로 변경한다
2. UserController 클래스에서 findUserByEmailAndPassword() 메서드명을 변경된 이름으로 수정한다

hyogenie-v committed:
refactor: UserRepository 인터페이스의 메서드 리턴값 변경

1. UserRepository 인터페이스의 findUser* 메서드가 Optional<User> 객체를 리턴하도록 수정한다

hyogenie-v committed:
LoginService에 대한 테스트 클래스 추가

hyogenie-v committed:
refactor: UserService 클래스의 updateUser 메서드가 수정된 User 객체를 리턴하도록 수정

hyogenie-v committed:
test: UserService에 대한 회원정보 수정 및 삭제 테스트 추가

1. UserRepository 객체를 MockBean으로 선언하여 테스트를 진행하도록 수정한다
2. 회원정보를 정상적으로 수정하는 경우에 대한 테스트 메서드를 추가한다
3. 회원이 탈퇴하는 경우에 대한 테스트 메서드를 추가한다

hyogenie-v committed:
test: 중복 코드 제거 & 테스트를 한 번에 돌리는 경우에 실패하는 원인 제거

1. consumeWith() 메서드를 returnResult() 메서드로 수정하여 동기적인 테스트로 변경한다
2. 각 테스트 메서드마다 사용하는 모든 객체가 유일하도록 수정한다
	2-1. 정적 변수인 Identifier를 선언하여 tearDown() 메서드에서 1씩 증가시킨다
	2-2. 엔티티 객체의 뒤에 Identifier를 붙인다
3. 중복된 어노테이션을 제거한다

hyogenie-v committed:
refactor: 컨트롤러에서 예외 발생 시 핸들링 방법 수정

1. 비정상적으로 로그인 및 로그아웃 경로로 접근했을 경우 예외를 처리하도록 수정한다
2. 예외가 발생했을 경우, 다른 URI로 리다이렉트하도록 변경한다

hyogenie-v committed:
test: LoginController에 대한 테스트 메서드 추가 및 리팩토링

hyogenie-v committed:
feat: UserController에 회원 탈퇴 시 로그인 여부 확인하는 로직 추가

hyogenie-v committed:
test: UserController에 대한 테스트 메서드 추가 및 리팩토링

gracefulBrown reviewed (approved):
피드백 반영하시느라 고생많으셨습니다~
테스트 코드 중복제거 하려는 노력 👍 
몇가지 코멘트 남기니 확인해주시고 다음 미션때 참고해주세요~
